src/chainstate/stacks/boot/mod.rs

```diff
@@ -124,7 +131,340 @@ pub fn make_contract_id(addr: &StacksAddress, name: &str) -> QualifiedContractId
+    /// Create an empty reward set where no one gets an early unlock
+    pub fn empty() -> RewardSet {
+        RewardSet {
+            rewarded_addresses: vec![],
+            start_cycle_state: PoxStartCycleInfo {
+                missed_reward_slots: vec![],
+            },
+        }
+    }
+}
+
 impl StacksChainState {
+    /// Return the MARF key used to store whether or not a given PoX
+    ///  cycle's "start" has been handled by the Stacks fork yet. This
+    ///  is used in Stacks 2.1 to help process unlocks.
+    fn handled_pox_cycle_start_key(cycle_number: u64) -> String {
+        format!("chainstate_pox::handled_cycle_start::{}", cycle_number)
+    }
+
+    /// Returns whether or not the `cycle_number` PoX cycle has been handled by the
+    ///  Stacks fork in the opened `clarity_db`.
+    pub fn handled_pox_cycle_start(clarity_db: &mut ClarityDatabase, cycle_number: u64) -> bool {
+        let db_key = Self::handled_pox_cycle_start_key(cycle_number);
+        match clarity_db.get::<String>(&db_key) {
+            Some(x) => x == POX_CYCLE_START_HANDLED_VALUE,
+            None => false,
+        }
+    }
+
+    fn mark_pox_cycle_handled(db: &mut ClarityDatabase, cycle_number: u64) {
+        let db_key = Self::handled_pox_cycle_start_key(cycle_number);
+        db.put(&db_key, &POX_CYCLE_START_HANDLED_VALUE.to_string());
+    }
+
+    /// Do all the necessary Clarity operations at the start of a PoX reward cycle.
+    /// Currently, this just means applying any auto-unlocks to Stackers who qualified.
+    pub fn handle_pox_cycle_start(
+        clarity: &mut ClarityTransactionConnection,
+        cycle_number: u64,
+        cycle_info: Option<PoxStartCycleInfo>,
+    ) -> Result<(), Error> {
+        clarity.with_clarity_db(|db| Ok(Self::mark_pox_cycle_handled(db, cycle_number)))?;
+
+        debug!("Handling PoX reward cycle start"; "reward_cycle" => cycle_number, "cycle_active" => cycle_info.is_some());
+
+        let cycle_info = match cycle_info {
+            Some(x) => x,
+            None => return Ok(()),
+        };
+
+        let pox_contract = boot::boot_code_id(POX_2_NAME, clarity.is_mainnet());
+
+        for (principal, amount_locked) in cycle_info.missed_reward_slots.iter() {
+            // we have to do several things for each principal
+            // 1. lookup their Stacks account and accelerate their unlock
+            // 2. remove the user's entries from every `reward-cycle-pox-address-list` they were in
+            //     (a) this can be done by moving the last entry to the now vacated spot,
+            //         and, if necessary, updating the associated `stacking-state` entry's pointer
+            //     (b) or, if they were the only entry in the list, then just deleting them from the list
+            // 3. correct the `reward-cycle-total-stacked` entry for every reward cycle they were in
+            // 4. delete the user's stacking-state entry.
+            clarity.with_clarity_db(|db| {
```

Comment:
This closure here scares me a little bit because (a) it's pretty big and (b) it's directly writing to and deleting contract state outside of a contract-call invocation (which I don't think we do anywhere else).

Could the bulk of this closure be lifted into a set of `(define-private ...)` functions within `.pox-2` that this closure could simply invoke?  We could easily verify that these functions are unreachable code from within the VM, and writing them in Clarity would (a) shorten the implementation somewhat, (b) let us re-use all of the VM code we currently rely on for mutating contract state, and (c) permit us to easily unit-test them.

What do you think?

Reply:
> We do something similar to this for the cost voting contract.

IIRC the Clarity VM will update a couple of Clarity DB key/values directly (`vm-costs::last-processed-at-height` and the cost-contract-specific `::state-summary` key), but -- correct me if I'm wrong -- it won't mutate state within the `cost-voting` contract's data space.  The VM uses the cost contract to determine what these key/values should be, in order to determine what the right cost functions are when it goes to evaluate a block.

Unrelated, looking into this caused me to notice a bug in the cost-voting system (https://github.com/stacks-network/stacks-blockchain/issues/3268) that we can fix for 2.1.

Reply:
Thanks for taking the time to do this in https://github.com/stacks-network/stacks-blockchain/pull/3260/commits/9d00a6cfd48fe557dc524a6935e7ab9c1af74d79.  I think the unlock implementation is a lot clearer and easier to understand now.